subsys/zigbee/osif/zb_nrf_serial.c

```diff
@@ -51,11 +48,142 @@ static size_t uart_tx_buf_len;
 static uint8_t uart_tx_buf_mem[CONFIG_ZIGBEE_UART_TX_BUF_LEN];
 
 static uint8_t *uart_rx_buf;
-static size_t uart_rx_buf_offset;
-static size_t uart_rx_buf_len;
+static volatile size_t uart_rx_buf_offset;
+static volatile size_t uart_rx_buf_len;
+
+/**
+ * Add (CONFIG_ZIGBEE_UART_RX_CHUNK_LEN - 1) to the buffer size, so the driver
+ * may write to the RX buffer directly, regardless the write pointer position.
+ */
+static uint8_t uart_rx_buf_mem[CONFIG_ZIGBEE_UART_RX_BUF_LEN +
+			       CONFIG_ZIGBEE_UART_RX_CHUNK_LEN - 1];
+static volatile size_t rx_write_i;
+static volatile size_t rx_req_i;
+static volatile size_t rx_read_i;
+
+/**
+ * Async: Push the buffer contents to the RX buffer.
+ *
+ * @param buf data buffer
+ * @param len data length
+ */
+static void rx_buf_push(uint8_t *buf, size_t len)
```

Comment:
In short: I looked at it and it was insufficient.
The missing feature, needed for this implementation was the possibility to write a constant-size data chunks at any time.
In the current Zephyr's ring buffer implementation you may plan the data reception through the `ring_buf_put_claim` API, but you cannot fix the RX size. The API returns the amount of data that can be written instead. I also hope that the ring buffer API is prepared to accept any amount of data, different from the claimed size, through the `ring_buf_put_finish` API call.

I looked at the implementation once more and found out that there is a small chance to use the existing implementation of the ring buffer, but it will require changes in UART ASYNC API to be nice and clean.
Currently we use the receive size to detect if the ASYNC UART has stopped the reception due to the RX timeout event. If it is the ring buffer that determines the RX size, either those values has to be stored (for both: ongoing and planned UART RX transactions) or the additional API for determining the timeout event has to be introduced.

What's more the transaction size defines the "blind spots", in which there is no chance to distinguish between regular RX and RX timeout event. If the size changes based on RX buffer contents, those "blind spots" became completely unpredictable.

I also do not see a point in spending much more time on refactoring the "ASYNC" version, so if you find it unacceptable in the current form - just let me know and I will squash those commits and remove this implementation from the history.